fpdf/html.py

```diff
@@ -656,36 +678,46 @@ def handle_starttag(self, tag, attrs):
                 size=tag_style.size_pt or self.font_size,
             )
             self.indent += 1
-            self._new_paragraph(top_margin=3, bottom_margin=3)
-            if self.tag_indents["blockquote"]:
-                self._write_paragraph("\u00a0" * self.tag_indents["blockquote"])
+            self._new_paragraph(
+                top_margin=3,
+                bottom_margin=3,
+                indent=self.tag_indents["blockquote"] * self.indent,
+            )
         if tag == "ul":
             self.indent += 1
             bullet_char = (
                 ul_prefix(attrs["type"]) if "type" in attrs else self.ul_bullet_char
             )
             self.bullet.append(bullet_char)
-            line_height = None
             if "line-height" in attrs:
                 try:
                     # YYY parse and convert non-float line_height values
-                    line_height = float(attrs.get("line-height"))
+                    self.line_height_stack.append(float(attrs.get("line-height")))
                 except ValueError:
                     pass
-            self._new_paragraph(line_height=line_height)
+            else:
+                self.line_height_stack.append(None)
+            if self.indent == 1:
+                self._new_paragraph(top_margin=self.list_top_margin, line_height=0)
```

Comment:
Applying the margins with the `</?[uo]l>` instead of the `<li>` is the correct and clean design. 👍 

But I'm not sure if checking for `self.indent == 1:` is the right criterion for the top margin.
What is the logic (or HTML spec) behind that?

Btw: 
You're using the NBSP to get the actual margin, with `list_top_margin` only adding a very small and probably unnecessary amount to it. If you do that, you could probably just leave `list_top_margin` away completely.

But then, actually using the `Paragraph()` top and bottom margin functionality has additional benefits. For example, the paragraph will not apply them at the top or bottom of a page. Your current solution will create an unnecessary empty space there, possibly resulting in unnecessary page breaks.

The better solution might be to just add a top/bottom margin of the current text size, and modify `Paragraph()` so it applies its margins even if it contains no text.

Reply:
> But I'm not sure if checking for self.indent == 1: is the right criterion for the top margin.
What is the logic (or HTML spec) behind that?

That is done to avoid applying margins in the case of nested lists, i.e. the margins are only applied if we aren't already handling an HTML element that increases `HTML2PDF.indent`.

> The better solution might be to just add a top/bottom margin of the current text size, and modify Paragraph() so it applies its margins even if it contains no text.

Will do.

Reply:
> That is done to avoid applying margins in the case of nested lists,

[The specs](https://html.spec.whatwg.org/multipage/rendering.html#lists) suggest the following defaults:

    dir, dl, menu, ol, ul { margin-block: 1em; }

which implies that nested lists also should have vertical margins (as there is no exception defined for them).

Browser makers seem to interpret that rather liberally, with both Firefox and Edge (hence Chromium) applying a margin above, but not below. They differ in that Firefox keeps the two nested bullets on the same line, while Chromium jumps one line:

![grafik](https://github.com/py-pdf/fpdf2/assets/17468844/1c55de1e-4900-4c76-9ad4-e9e1d943e11a)  ![grafik](https://github.com/py-pdf/fpdf2/assets/17468844/060d7422-f152-4752-ae77-730095c0ef47)

I guess only using vertical margins for the top level list is just as compliant as those two...